### ISSUE TITLE
Direction of go-critic

### ISSUE THREAD

**Issue by dominikh** [go-critic/go-critic#562]
I am wondering about the direction that go-critic is taking. Quoting the README,

> Go source code linter providing checks currently missing from other linters.
> 
> When a specific check implementation is mature and proven useful, propose its integration into other linters. If a good checker can't find a better home, it stays here

it would seem to me that you wouldn't spend a large portion of time on implementing checks that already exist in other linters. Yet you're on your way to doing exactly that.

busySelect, dupSubExpr, #560, #516, #413 and #291 all already exist in other checkers, or are minor improvements on these existing checks.

What gives?

**Comment by Quasilyte** [comment]
Usually, we're running `gomentalinter --enable-all` as a quick check whether something does detect proposed problem. If nothing does, it's hard to determine whether there is a partial implementation somewhere or is it some kind of a corner case that is not handled for some reason. There were other commands, but I unfortunately don't remember them right now.

For example, staticcheck does detect some condition expression duplication, but not ones that report such code as suspicious:
```go
xs[i] < xs[i]
```
I would open an issue on staticcheck if that check is stabilized and had 0 false positives, but otherwise, it's really out of my area of comfort to do frequent reports, I want to get some fun for things I do for free in my spare time. One of the problems is that there is very little time dedicated for real testing, so most of our experimental checks can be broken in a different ways.

`unusedParam` is not related to `unused` at all. It only suggest name changes for `xyz` to `_`.

And for the `lenNil`, it seems that we get it wrong and all useful cases are already covered by staticcheck. And we would not understand this before we had an implementation that makes it possible to find false positives. If that extension would be useful for staticcheck, it can implement that on it's own, granted that we already did testing on real projects, found some corner cases, etc.

I'm not the author of other PRs, but before implementation is merged, I do check whether it's already available somewhere.

Please, also note that most active maintainers use this one as a place to learn how to write static analysis tools. I'm using gocritic to do other open source projects review and send fixes to them (already found 2 bugs in `Less` methods that used duplicated index, so it was useful to implement

**Comment by dominikh** [comment]
Please don't get me wrong. Everybody can write whatever code they want. Play, experiment, have fun! I just wanted to make sure that you (and your users) were aware of the overlap with existing tools.

P.S.:

> For example, staticcheck does detect some condition expression duplication, but not ones that report such code as suspicious:

```
xs[i] < xs[i]
```

Oh, but it does:

```
desktop ~ $ cat /tmp/foo.go
package pkg

func fn(xs []string, i int) {
	if xs[i] < xs[i] {
		println(true)
	}
}

desktop 1? ~ $ staticcheck /tmp/foo.go
/tmp/foo.go:4:5: identical expressions on the left and right side of the '<' operator (SA4000)
```

**Comment by Quasilyte** [comment]
Is that some recent extension?

I've tried on this code:
```go
func fn(xs []string, i int) {
	return xs[i] < xs[i] 
}
```
Through gometalinter, it gave no warnings. Maybe it vendors older staticcheck?

**Comment by Quasilyte** [comment]
> I just wanted to make sure that you (and your users) were aware of the overlap with existing tools.

I'm trying to verify beforehand. Maybe there is a need for a good verification script that not only runs `gometalinter --enable`, but also most well-known linters built from tip branch.

**Comment by dominikh** [comment]
That check has been around since `Sat May 7 01:06:16 2016 +0200`, and gometalinter does catch it for me.

**Comment by Quasilyte** [comment]
OK, now I can reproduce that too:
```
$ cat src/foo/foo.go 
package foo

type foo struct{ xs []string }

func (foo foo) Less(i, j int) bool {
	return foo.xs[i] < foo.xs[i]
}

$ gometalinter --enable-all ./src/foo
src/foo/foo.go:6:9:warning: identical expressions on the left and right side of the '<' operator (SA4000) (megacheck)
```

Don't remember what test I've initially used to check whether it detected by something from gometalinter.

**Comment by dominikh** [comment]
Well, I do hope your implementation doesn't flag `x + x`, that'd be a false positive ;)

At any rate, I'm not sure there's anything left to do in this issue, so feel free to close it.

**Comment by Quasilyte** [comment]
OK. Hope I've covered some details you was trying to find.

I think neither of linters will get "all checks possible in the world".
Personally, I'll try to concentrate on the checks that are not implemented yet, and if they feel right, I can contact you to share the idea. I'm not opposed to moving checks from gocritic to staticcheck. Reduces the amount of maintenance required while we still have same amount of checks in total.

But note that my opinion may be alien for some maintainers, but hopefully we can get over some human-related factors and concentrate on the "bringing more static checks" part.
